web/packages/shared/hooks/useAsync.ts

```diff
@@ -163,13 +164,43 @@ export class CanceledError extends Error {
   }
 }
 
-export type AttemptStatus = 'processing' | 'success' | 'error' | '';
+export type AttemptStatus = Attempt<any>['status'];
 
-export type Attempt<T> = {
-  data?: T;
-  status: AttemptStatus;
-  statusText: string;
-};
+export type Attempt<T> =
+  | {
+      status: '';
+      data: null;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'processing';
+      /** data is either null or contains data from the previous success attempt if the attempt was retried. */
+      data: null | T;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'success';
+      data: T;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'error';
+      data: null;
+      statusText: string;
+      error: any;
+    };
```

Comment:
Representing `Attempt` like this helps with understanding which fields are available when. `processing` shouldn't have access to the `error` field and so on.

Reply:
I think this is good. I was trying to think of a scenario where we'd want to know the previous error while processing another request but now that I think about it, errors are always removed when trying the next request as to not confuse stale errors so, yeah this is great!

Reply:
Regarding stale values, in most implementations of a similar data structure that I'm familiar with (bringing up RemoteData from Elm again as the golden standard), it's not even possible to read stale _data_ when the action is being retried and is in progress.

In our implementation, this is totally possible as `data` gets overriden only on success or error.

---

Your comment also made me realize that we override `error` only if the callback rejects, so in theory it's possible to see something like `{status: 'success', data: /* something */, error: Error, statusText: 'some error message'}`. However, TS would not allow you to access `error` in that case anyway. And changing this behavior could technically introduce a regression so I guess I'm not going to touch it for now.